apiserver/facades/controller/caasoperatorprovisioner/provisioner.go

```diff
@@ -57,3 +68,159 @@ func (a *API) WatchApplications() (params.StringsWatchResult, error) {
 	}
 	return params.StringsWatchResult{}, watcher.EnsureErr(watch)
 }
+
+// UpdateApplicationsUnits updates the Juju data model to reflect the given
+// units of the specified application.
+func (a *API) UpdateApplicationsUnits(args params.UpdateApplicationUnitArgs) (params.ErrorResults, error) {
+	result := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Args)),
+	}
+	if len(args.Args) == 0 {
+		return result, nil
+	}
+	for i, appUpdate := range args.Args {
+		appTag, err := names.ParseApplicationTag(appUpdate.ApplicationTag)
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		app, err := a.state.Application(appTag.Id())
+		if err != nil {
+			result.Results[i].Error = common.ServerError(err)
+			continue
+		}
+		if err := a.updateUnitsFromCloud(app, appUpdate.Units); err != nil {
+			result.Results[i].Error = common.ServerError(err)
+		}
+	}
+	return result, nil
+}
+
+// updateUnitsFromCloud takes a slice of unit information provided by an external
+// source (typically a cloud update event) and merges that with the existing unit
+// data model in state. The passed in units are the complete set for the cloud, so
+// any existing units in state with provider ids which aren't in the set will be removed.
+func (a *API) updateUnitsFromCloud(app Application, units []params.ApplicationUnitParams) error {
+	// Set up the initial data structures.
+	existingStateUnits, err := app.AllUnits()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	var (
+		// unassociatedUnits are those which do not yet
+		// have a provider id assigned.
+		unassociatedUnits []Unit
+
+		// removedUnits are those that exist in state but are not
+		// represented in the unit params passed in.
+		removedUnits []Unit
+
+		addedCloudUnits []params.ApplicationUnitParams
+	)
+	stateUnitsById := make(map[string]Unit)
+	cloudUnitIds := make(set.Strings)
+
+	existingCloudUnitsById := make(map[string]params.ApplicationUnitParams)
+
+	// Record all unit provider ids known to exist in the cloud.
+	for _, u := range units {
+		cloudUnitIds.Add(u.Id)
+	}
+
+	// Loop over any existing state units and record those which do not yet have
+	// provider ids, and those which have been removed or updated.
+	for _, u := range existingStateUnits {
+		if u.Life() != state.Alive {
+			continue
```

Comment:
This doesn't look to be any better? what I was saying is that if you have a unit in state that's Dying, but has the provider ID, surely we shouldn't go and associate the pod with a *new* unit. The unit dying in state should cause that pod to be destroyed? Maybe things don't work how I expect? (if so, more comments)

Reply:
No, I think you are right. I've reworked a bit.

Reply:
thanks, that looks better now